src/main/java/io/reactivex/internal/operators/flowable/FlowableThrottleAndSample.java

```diff
@@ -0,0 +1,194 @@
+import java.util.concurrent.atomic.AtomicReference;
+
+public final class FlowableThrottleAndSample<T> extends AbstractFlowableWithUpstream<T, T> {
+
+    private final long windowDuration;
+    private final TimeUnit unit;
+    private final Scheduler scheduler;
+
+    private final boolean emitLast;
+
+    public FlowableThrottleAndSample(@NonNull final Flowable<T> source, final long windowDuration,
+                                     final TimeUnit unit, final Scheduler scheduler, final boolean emitLast) {
+        super(source);
+        this.windowDuration = windowDuration;
+        this.unit = unit;
+        this.scheduler = scheduler;
+        this.emitLast = emitLast;
+    }
+
+    @Override
+    protected void subscribeActual(final Subscriber<? super T> s) {
+        final SerializedSubscriber<T> serial = new SerializedSubscriber<T>(s);
+        if (emitLast) {
+            source.subscribe(new ThrottleAndSampleEmitLast<T>(serial, windowDuration, unit, scheduler.createWorker()));
+        } else {
+            source.subscribe(new ThrottleAndSampleNoLast<T>(serial, windowDuration, unit, scheduler.createWorker()));
+        }
+    }
+
+    abstract static class ThrottleAndSampleSubscriber<T> extends AtomicReference<T> implements FlowableSubscriber<T>, Subscription, Runnable {
+
+        private static final long serialVersionUID = -7130465637537281443L;
+
+        protected final Subscriber<? super T> actual;
+        private final long windowDuration;
+        private final TimeUnit unit;
+        private final Scheduler.Worker worker;
+
+        final AtomicLong requested = new AtomicLong();
+
+        private final SequentialDisposable timer = new SequentialDisposable(Disposables.disposed());
+
+        private Subscription subscription;
+
+        boolean isDone;
+
+        ThrottleAndSampleSubscriber(final Subscriber<? super T> actual, long windowDuration,
+                                    final TimeUnit unit, final Scheduler.Worker worker) {
+            this.actual = actual;
+            this.windowDuration = windowDuration;
+            this.unit = unit;
+            this.worker = worker;
+        }
+
+        @Override
+        public void onSubscribe(final Subscription subscription) {
+            if (SubscriptionHelper.validate(this.subscription, subscription)) {
+                this.subscription = subscription;
+                actual.onSubscribe(this);
+                subscription.request(Long.MAX_VALUE);
+            }
+        }
+
+        @Override
+        public void onNext(final T value) {
+            final Disposable throttling = timer.get();
+            if (throttling.isDisposed()) {
+                emit(value);
+                timer.replace(worker.schedule(this, windowDuration, unit));
+            } else {
+                set(value);
+            }
+        }
+
+        @Override
+        public void run() {
+            final Disposable throttling = timer.get();
+            throttling.dispose();
```

Comment:
There is a race here. If `run()` gets here and `onNext` is fired, throttling will be disposed and the `onNext` value gets emitted. Here then the cached value gets emitted as well and now there are two tasks delayed for the subsequent interactions.